README.md
Exchange arbitration data collection worker - Bittrex
=====================================================

[![build status](https://gitlab.com/x-data/x-financial-arbitration-data-collection-worker-bittrex/badges/master/build.svg)](https://gitlab.com/x-data/x-financial-arbitration-data-collection-worker-bittrex/commits/master)
[![coverage](https://gitlab.com/x-data/x-financial-arbitration-data-collection-worker-bittrex/badges/master/coverage.svg?job=test)](https://gitlab.com/x-data/x-financial-arbitration-data-collection-worker-bittrex/commits/master)

This worker aims to collect data from the exchange Bittrex.

**NOTE:** The `price` field should always be in the `quote asset` value

**NOTE:** This project relies on environment variables to work, so make sure to set up everything before running.

## Running the tests

Requirements:

    * Node (tested with 8.9.4)
    * npm (tested with (5.6.0)

Run:

    npm test

Done. It will check lint issues and run the available tests, displaying a coverage status.

## Installing and using

Here you will find two section, running with docker and without it. We definitely recommend the docker usage for it's simplicity and other several benefits

### Running with docker

Requirements:

    * Docker (tested with 17.12.0-ce)
    * Docker Compose (tested with 1.18.0)

Running the project:

    docker-compose up

### Running without docker

Requirements:

    * Node (tested with 8.9.4)
    * npm (tested with (5.6.0)

Installing dependencies:

    npm install

Running the project:

    npm run local

Environment variables available are listed below and you can check it's usage in the `src/config/default.js` file.

| environment variable    | default                                       | required | description |
|-------------------------|-----------------------------------------------|----------|-------------|
| EXCHANGE_NAME           | bittrex                                       | true     | it becomes *required* when the there is no *SYMBOL_DELIMTER* separating the pair of currencies in the exchange |
| SYMBOL_DELIMITER        |                                               | false    | it defines the pair of currencies separator, e.g. BTC-LTC the `-` is the separator |
| API_URL                 | https://bittrex.com/api/v1.1/public/getticker | true     | the full API URL route/path you can collect the pair prices |
| COLLECT_ONLY_MATCHES    | false                                       | false    | if you wish to collect only symbols that match with other exchanges |
| COLLECT_SPECIFIC_PRICES | true                                          | false    | if you wish to collect only specific prices and not all the prices available in the exchange |
| COLLECT_PAIRS           | ETH-BTC,BCH-BTC                               | false    | the comma separated list of pairs to collect, e.g. "BTC-LTC, ETH-XRP" |
| INTERVAL_IN_SECONDS     | 300                                           | false    | the interval in seconds to update the pairs |
| NODE_CONFIG_DIR         |                                               | true     | set it to `./src/config` |
| RUNNING_MODE            | single-mode                                   | false    | it specifies if the collect data process should run once (`single-mode`) or forever as a `worker` |
| DB_NAME                 | db                                            | true     | the database name |
| DB_USERNAME             | worker                                        | true     | the username that will login into the database |
| DB_PASSWORD             | worker                                        | true     | the database password |
| DB_HOST                 | localhost                                     | true     | the database host |
| DB_PORT                 | 3306                                          | true     | the database port |
| DB_DEBUG                | false                                         | true     | debugging the database |
| DB_LOGGING              | undefined                                     | true     | whether the database should log or not |
| DB_POOL_MAX             | 20                                            | true     | the maximum pool number |
| DB_POOL_MIN             | 5                                             | true     | the minimum pool number |
| DB_POOL_IDLE            | 10000                                         | true     | the number of pools allowed to be idle |
| LOGGING_LEVEL           | debug                                         | true     | the application logging level |


CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
* get only matches

### Updated
* migrate project from gitlab to github
* update project name
* update node/npm versions

### Fixed
* audit fix in packages

## [1.0.0] - 2018-03-02
### Added
* blacklist of invalid pairs
* mapped different host port to database
* worker


src/worker.spec.js
const assert = require('chai').assert
const sinon = require('sinon')
const axios = require('axios')
const config = require('config')

const Worker = require('./worker')
const Data = require('./utils/sequelize').Data

describe('worker', () => {
  const stubRequest = sinon.stub(axios, 'get')
  const stubData = sinon.stub(Data, 'create')

  beforeEach(() => {
    stubRequest.reset()
    stubData.reset()
  })

  after(() => {
    stubRequest.restore()
    stubData.restore()
  })

  it('get data', (done) => {
    const normalizedSymbol = 'BCH-BTC'
    const denormalizedSymbol = 'BCH-BTC'

    stubRequest.resolves({data: {result: {'Last': 1.01230}, 'success': true}})

    Worker.getData(normalizedSymbol).then((response) => {
      assert.strictEqual(stubRequest.calledOnce, true, 'request was called once')
      assert.strictEqual(
        stubRequest.calledWith(`${config.api.url}?market=${denormalizedSymbol}`),
        true,
        'called with API URL and symbol'
      )

      done()
    })
  })

  it('start worker in single-mode', (done) => {
    const denormalizedSymbol = 'BCH-BTC'
    console.log(config.collect.pairs)
    console.log(config.collect.pairs.split(',').length)
    const numberOfPairs = config.collect.pairs.split(',').length

    stubRequest.resolves({data: {result: {'Last': 1.01230}, success: true}})

    Worker.start().then(() => {
      console.log('call count', stubData.callCount)
      assert.strictEqual(stubData.callCount, numberOfPairs, 'saved data for each pair')

      assert.strictEqual(stubRequest.callCount, numberOfPairs, 'request was called for each pair')
      assert.strictEqual(
        stubRequest.calledWith(`${config.api.url}?market=${denormalizedSymbol}`),
        true,
        'called with API URL and symbol'
      )

      done()
    })
  })
})


src/worker.js
const R = require('ramda')
const axios = require('axios')
const config = require('config')
const normalize = require('x-cryptocurrencies-normalizr')
const logger = require('./utils/logging').logger
const Data = require('./utils/sequelize').Data

const Worker = () => {}

Worker.blackList = []

Worker.getData = async (normalizedPair) => {
  /*
   * The `normalizedPair` parameter is normalized accross all workers
   * since we can't send the request with that value
   * we must first denormalize it
  * */

  let url = config.api.url
  let exchangePair
  if (normalizedPair) {
    exchangePair = normalize.denormalize.pair(normalizedPair, config.get('exchange').name)
    url += `?market=${exchangePair}`
  }

  const normalizedPairList = normalizedPair.split('-')

  return axios.get(url)
    .then((res) => {
      if (!res.data.success) {
        logger.error(res.data)

        if (res.data.message === 'INVALID_MARKET') {
          Worker.blackList.push(normalizedPair)
        }

        return undefined
      }

      return {
        baseAsset: normalizedPairList[0],
        quoteAsset: normalizedPairList[1],
        normalizedPair: normalizedPair,
        denormalizedPair: exchangePair,
        price: res.data.result.Last.toString()
      }
    })
    .catch((err) => {
      logger.error(
        `error requesting ${url}.
         Error: ${err}`
      )
    })
}

Worker.call = async () => {
  const requestList = []

  if (config.collect.onlyMatches === true) {
    let pairs = normalize.findMatches()

    for (let i = 0; i < pairs.length; i++) {
      let pair = pairs[i]

      if (R.indexOf(pair, Worker.blackList) > -1) {
        continue
      }

      requestList.push(
        Worker.getData(pair)
      )
    }
  } else if (config.collect.specificPrices === false) {
    // collect all prices available
    requestList.push(
      Worker.getData()
    )
  } else {
    // collect only a given set of pairs
    if (config.collect.pairs === undefined) {
      throw new Error(
        `in order to collect specific prices you must set \`COLLECT_PAIRS\` with comma separated pairs.
         e.g.: export COLLECT_PAIRS=BTC-ETH,BTC-IOTA,ETH-IOTA
        `
      )
    }

    const coinPairs = R.split(',', config.collect.pairs)

    for (let i = 0; i < coinPairs.length; i++) {
      let coinPair = coinPairs[i]

      if (R.indexOf(coinPair, Worker.blackList) > -1) {
        continue
      }

      requestList.push(
        Worker.getData(coinPair)
      )
    }
  }

  return Promise.all(requestList)
    .then(async (responses) => {
      for (let i = 0; i < responses.length; i++) {
        let response = responses[i]
        if (!response) {
          continue
        }

        logger.debug(response)
        await Data.create(response)
      }
    })
    .catch((err) => {
      logger.error(err)
    })
}

Worker.start = () => {
  logger.info(`starting with ${config.running.mode} mode and interval set to ${config.interval}`)

  return new Promise((resolve, reject) => {
    if (config.running.mode === 'single-time' || config.interval === undefined) {
      Worker.call()
        .then(() => resolve())
        .catch((err) => {
          logger.error(err)
          return reject(err)
        })
      return
    }

    setInterval(Worker.call, config.interval)
    return resolve()
  })
}

module.exports = Worker


src/config/default.js
const path = require('path')
const env = process.env.NODE_ENV || 'development'
const root = path.join(__dirname, '..', '..')

try {
  require('dotenv').config({path: path.join(root, `.env-${env.toLowerCase()}`)})
} catch (err) {
  console.log(err)
}

const base = {
  exchange: {
    name: process.env.EXCHANGE_NAME || 'bittrex',
    symbol: {
      delimiter: process.env.SYMBOL_DELIMITER
    }
  },
  api: {
    url: process.env.API_URL || 'https://bittrex.com/api/v1.1/public/getticker'
  },
  collect: {
    specificPrices: process.env.COLLECT_SPECIFIC_PRICES !== 'false',
    onlyMatches: process.env.COLLECT_ONLY_MATCHES === 'true',
    pairs: process.env.COLLECT_PAIRS || 'ETH-BTC,BCH-BTC,BTC-LTC,BTC-ETH,BTC-XRP,BTC-NEO'
  },
  interval: parseFloat(process.env.INTERVAL_IN_SECONDS || 300) * 1000,
  running: {
    mode: process.env.RUNNING_MODE || 'single-time'
  },
  db: {
    knex: {
      dialect: 'mysql2',
      tablename: 'knex_migrations'
    },
    name: process.env.DB_NAME || 'db',
    username: process.env.DB_USERNAME || 'worker',
    password: process.env.DB_PASSWORD || 'worker',
    options: {
      dialect: 'mysql',
      host: process.env.DB_HOST || 'localhost',
      port: process.env.DB_PORT || 3306,
      debug: process.env.DB_DEBUG || false,
      logging: process.env.DB_LOGGING || undefined,
      pool: {
        max: process.env.DB_POOL_MAX || 20,
        min: process.env.DB_POOL_MIN || 5,
        idle: process.env.DB_POOL_IDLE || 10000
      }
    }
  },
  logging: {
    level: process.env.LOGGING_LEVEL || 'debug',
    project: {
      name: 'x-financial-arbitration-data-collection-worker-bittrex'
    }
  }
}

module.exports = base
